File: apps/api/MyPetLink.Api/Auth/AdminCapabilities.cs
namespace MyPetLink.Api.Auth;

/// <summary>
/// The Admin Portal capability catalogue — the single authority for what an
/// operator may do. Every protected admin endpoint names one of these, every
/// role stores a subset of them, and the Admin Portal renders its navigation
/// and action buttons from the set the API reports back.
///
/// Adding a capability here is the only supported way to create one. A string
/// that is not in this catalogue is rejected when a role is saved and ignored
/// when access is resolved, so no request payload can invent permission.
/// </summary>
public static class AdminCapabilities
{
    // --- Access management ----------------------------------------------------
    public const string AdminUsersView = "admin.users.view";
    public const string AdminUsersManage = "admin.users.manage";
    public const string AdminRolesView = "admin.roles.view";
    public const string AdminRolesManage = "admin.roles.manage";
    public const string AuditLogView = "audit_log.view";

    // --- Retail orders --------------------------------------------------------
    public const string OrdersView = "orders.view";
    public const string OrdersExport = "orders.export";
    public const string OrdersManage = "orders.manage";
    public const string OrdersShippingManage = "orders.shipping.manage";
    public const string OrdersTagsAssign = "orders.tags.assign";

    // --- Payment proofs -------------------------------------------------------
    public const string PaymentProofsView = "payment_proofs.view";
    public const string PaymentProofsExport = "payment_proofs.export";
    public const string PaymentProofsReview = "payment_proofs.review";

    // --- Tag inventory --------------------------------------------------------
    public const string InventoryView = "inventory.view";
    public const string InventoryExport = "inventory.export";
    public const string InventoryManage = "inventory.manage";
    public const string InventoryGenerate = "inventory.generate";
    public const string InventoryReceiptsManage = "inventory.receipts.manage";
    public const string InventoryCostsView = "inventory.costs.view";

    // --- Smart Tags -----------------------------------------------------------
    public const string SmartTagsView = "smart_tags.view";
    public const string SmartTagsExport = "smart_tags.export";
    public const string SmartTagsManage = "smart_tags.manage";
    public const string SmartTagsAssign = "smart_tags.assign";
    public const string SmartTagsTransfer = "smart_tags.transfer";

    // --- Tag catalog ----------------------------------------------------------
    public const string CatalogView = "catalog.view";
    public const string CatalogManage = "catalog.manage";

    // --- Customers ------------------------------------------------------------
    public const string OwnersView = "owners.view";
    public const string OwnersManage = "owners.manage";
    public const string OwnersExport = "owners.export";
    public const string PetsView = "pets.view";
    public const string PetsManage = "pets.manage";
    public const string PetsExport = "pets.export";

    // --- Merchant sales -------------------------------------------------------
    // "Reseller" and "merchant" are the same record in this system, so sales.*
    // deliberately governs both rather than adding a parallel resellers.* key.
    public const string SalesView = "sales.view";
    public const string SalesManage = "sales.manage";
    public const string MerchantOrdersView = "merchant_orders.view";
    public const string MerchantOrdersManage = "merchant_orders.manage";
    public const string MerchantOrdersFulfil = "merchant_orders.fulfil";
    public const string MerchantDocumentsSend = "merchant_documents.send";

    // --- Finance --------------------------------------------------------------
    public const string MerchantInvoicesView = "merchant_invoices.view";
    public const string MerchantInvoicesManage = "merchant_invoices.manage";
    public const string MerchantInvoicesRecordPayment = "merchant_invoices.record_payment";
    public const string SalesCommissionsView = "sales_commissions.view";
    public const string SalesCommissionsExport = "sales_commissions.export";
    public const string SalesCommissionsReverse = "sales_commissions.reverse";
    public const string SalesCommissionRulesManage = "sales_commissions.rules.manage";
    public const string PayoutsView = "payouts.view";
    public const string PayoutsManage = "payouts.manage";
    public const string PayoutsSettle = "payouts.settle";

    // --- Marketing ------------------------------------------------------------
    public const string MarketingView = "marketing.view";
    public const string MarketingManage = "marketing.manage";

    // --- Configuration --------------------------------------------------------
    public const string PlansView = "plans.view";
    public const string SettingsView = "settings.view";
    public const string SettingsManage = "settings.manage";
    public const string EmailTemplatesView = "email_templates.view";
    public const string EmailTemplatesManage = "email_templates.manage";
    public const string SampleExperienceView = "sample_experience.view";
    public const string SampleExperienceManage = "sample_experience.manage";

    // --- System ---------------------------------------------------------------
    public const string OperationalStatusView = "operational_status.view";
}

File: apps/web/src/lib/adminCapabilities.ts
// Admin Portal capability keys, mirroring the API's capability catalogue.
//
// These names exist so the portal can decide what to show. They are NOT the
// authorization: every protected destination and action is enforced by the API
// against the same key, so calling it directly without the capability is
// refused. Hiding a button here is a courtesy to the operator, never a control.

export const adminCapabilities = {
  // Access management
  adminUsersView: "admin.users.view",
  adminUsersManage: "admin.users.manage",
  adminRolesView: "admin.roles.view",
  adminRolesManage: "admin.roles.manage",
  auditLogView: "audit_log.view",

  // Retail orders
  ordersView: "orders.view",
  ordersExport: "orders.export",
  ordersManage: "orders.manage",
  ordersShippingManage: "orders.shipping.manage",
  ordersTagsAssign: "orders.tags.assign",

  // Payment proofs
  paymentProofsView: "payment_proofs.view",
  paymentProofsExport: "payment_proofs.export",
  paymentProofsReview: "payment_proofs.review",

  // Tag inventory
  inventoryView: "inventory.view",
  inventoryExport: "inventory.export",
  inventoryManage: "inventory.manage",
  inventoryGenerate: "inventory.generate",
  inventoryReceiptsManage: "inventory.receipts.manage",
  inventoryCostsView: "inventory.costs.view",

  // Smart Tags
  smartTagsView: "smart_tags.view",
  smartTagsExport: "smart_tags.export",
  smartTagsManage: "smart_tags.manage",
  smartTagsAssign: "smart_tags.assign",
  smartTagsTransfer: "smart_tags.transfer",

  // Tag catalog
  catalogView: "catalog.view",
  catalogManage: "catalog.manage",

  // Owners and pets
  ownersView: "owners.view",
  ownersManage: "owners.manage",
  ownersExport: "owners.export",
  petsView: "pets.view",
  petsManage: "pets.manage",
  petsExport: "pets.export",
  ownerSocialHandleAssign: "owners.social_handle.assign",

  // Community moderation
  communityReportsView: "community_reports.view",
  communityReportsResolve: "community_reports.resolve",
  communityModerationEnforce: "community_moderation.enforce",

  // Merchant sales
  salesView: "sales.view",
  salesManage: "sales.manage",
  merchantOrdersView: "merchant_orders.view",
  merchantOrdersManage: "merchant_orders.manage",
  merchantOrdersFulfil: "merchant_orders.fulfil",
  merchantDocumentsSend: "merchant_documents.send",

  // Finance
  merchantInvoicesView: "merchant_invoices.view",
  merchantInvoicesManage: "merchant_invoices.manage",
  merchantInvoicesRecordPayment: "merchant_invoices.record_payment",
  salesCommissionsView: "sales_commissions.view",
  salesCommissionsExport: "sales_commissions.export",
  salesCommissionsReverse: "sales_commissions.reverse",
  salesCommissionRulesManage: "sales_commissions.rules.manage",
  payoutsView: "payouts.view",
  payoutsManage: "payouts.manage",
  payoutsSettle: "payouts.settle",

  // Marketing
  marketingView: "marketing.view",
  marketingManage: "marketing.manage",

  // Configuration
  plansView: "plans.view",
  settingsView: "settings.view",
  settingsManage: "settings.manage",
  emailTemplatesView: "email_templates.view",
  emailTemplatesManage: "email_templates.manage",
  sampleExperienceView: "sample_experience.view",
  sampleExperienceManage: "sample_experience.manage",

  // System
  operationalStatusView: "operational_status.view",
} as const;

export type AdminCapabilityKey =
  (typeof adminCapabilities)[keyof typeof adminCapabilities];

/**
 * What the signed-in operator may do, as reported by the API.
 *
 * `granted` is already expanded: someone whose role grants everything arrives
 * here with every key listed, so nothing downstream needs a special case.
 */
export type AdminAccessCapabilities = {
  isSuperAdmin: boolean;
  roles: readonly AdminAssignedRole[];
  granted: ReadonlySet<string>;
};

export type AdminAssignedRole = {
  roleId: string;
  code: string;
  name: string;
  grantsAllCapabilities: boolean;
};

/** Deny by default. This is what an unresolved or refused session looks like. */
export const noAdminAccess: AdminAccessCapabilities = {
  isSuperAdmin: false,
  roles: [],
  granted: new Set<string>(),
};

/** Every capability, for local preview where there is no API to ask. */
export function allAdminCapabilities(): AdminAccessCapabilities {
  return {
    isSuperAdmin: true,
    roles: [],
    granted: new Set<string>(Object.values(adminCapabilities)),
  };
}

export function hasCapability(
  access: AdminAccessCapabilities,
  capability: AdminCapabilityKey
): boolean {
  return access.granted.has(capability);
}

export function hasAnyCapability(
  access: AdminAccessCapabilities,
  capabilities: readonly AdminCapabilityKey[]
): boolean {
  return capabilities.length === 0 || capabilities.some((key) => access.granted.has(key));
}

export function hasEveryCapability(
  access: AdminAccessCapabilities,
  capabilities: readonly AdminCapabilityKey[]
): boolean {
  return capabilities.every((key) => access.granted.has(key));
}

export function toAdminAccessCapabilities(
  source:
    | {
        isSuperAdmin?: boolean;
        roles?: readonly AdminAssignedRole[];
        capabilities?: readonly string[];
      }
    | null
    | undefined
): AdminAccessCapabilities {
  if (!source) {
    return noAdminAccess;
  }

  return {
    isSuperAdmin: Boolean(source.isSuperAdmin),
    roles: source.roles ?? [],
    granted: new Set<string>(source.capabilities ?? []),
  };
}
